components/constellation/constellation.rs

```diff
@@ -204,6 +204,20 @@ pub struct Constellation<Message, LTF, STF> {
     /// The last frame tree sent to WebRender.
     active_browser_id: Option<TopLevelBrowsingContextId>,
 
+    /// All the browsers in constellation. Since the active browser
+    /// (`active_browser_id`) may change with the `SelectBrowser` message (e.g.
+    /// when user switches to another tab), we must keep track of each browser's
+    /// focused browsing context so that we can restore the
+    /// `focused_browsing_context_id` to point to the active browser's focused
+    /// browsing context.
+    browser_ids: HashMap<TopLevelBrowsingContextId, BrowsingContextId>,
+
+    /// The currently focused browsing context for key events. The focused
+    /// pipeline is the current entry of the focused browsing context. This
+    /// field is always updated to be the focused browsing context of the
+    /// current active browser.
+    focused_browsing_context_id: Option<BrowsingContextId>,
```

Comment:
This could store the active/focused top level browsing context: `TopLevelBrowsingContextId`. With `focused_browsing_contexts` you can use this as the key and retrieve the focused `BrowsingContextId` for key events to go to. (This will all need to be reworked with multi-window support, but for now this will do).

Possibly we could just reuse `active_browser_id` as it seems like it would be the equivalent.

Reply:
I was thinking about this too, but I left it in because I wasn't sure of the additional overhead that always looking up the related browsing context in the hash map would introduce in `handle_key_events` (since from what I understand this might be a hot path). That is, I viewed `focused_browsing_context` as a sort of cache. Do you think this is valid or just premature optimization?

Reply:
But I agree that it makes the logic more confusing, so it's perhaps worth removing either way.

Reply:
Right, doing an extra lookup would add a little extra overhead here. I don't imagine it will be noticeable though (if I am wrong, we can switch back to this caching method).